File: 3296-minimum-number-of-seconds-to-make-mountain-height-zero/3296-minimum-number-of-seconds-to-make-mountain-height-zero.py
import math
class Solution(object):
    def minNumberOfSeconds(self, h, workerTimes):
        """
        :type mountainHeight: int
        :type workerTimes: List[int]
        :rtype: int
        """
        mintime=1
        base=workerTimes[0]
        maxtime=base*(h*(h+1))//2
        ans=maxtime
        while maxtime>=mintime:
            mid=(maxtime+mintime)//2
            height=0
            for i in workerTimes:
                z=mid/i
                if z>0:
                    count=(math.sqrt(1+8*z)-1)//2
                    height+=count
            if height<h:
                mintime=mid+1
            else:
                ans=mid
                maxtime=mid-1
        return ans 
            

File: 3296-minimum-number-of-seconds-to-make-mountain-height-zero/3296-minimum-number-of-seconds-to-make-mountain-height-zero.cpp
class Solution {
public:
typedef long long ll;
ll bs(ll mid,ll val){
     ll z=mid/val;
     z;
     ll start=1;
     ll end=z;
     ll v=end;
     while(start<=end){
        ll m=start+(end-start)/2;
        if((__int128)m*(m+1)/2<=z){
            v=m;
            start=m+1;
        }
        else{
            end=m-1;
        }
     }
     return v;
}
bool check(int mountainHeight, vector<int>& workerTimes,ll mid){
    ll total=0;
    int n=workerTimes.size();
    for(int i=0;i<n;i++){
       ll count=bs(mid,workerTimes[i]);
       total+=count;
    }
    if(total>=mountainHeight)return true;
    return false;
}
    long long minNumberOfSeconds(int mountainHeight, vector<int>& workerTimes) {
        long long l=0 ,r=LLONG_MAX;
        long long ans=r;
        while(l<=r){
            ll mid=l+(r-l)/2;
            if(check(mountainHeight,workerTimes,mid)){
                ans=mid;
                r=mid-1;
            }
            else{
                l=mid+1;
            }
        }
        return ans;

    }
};